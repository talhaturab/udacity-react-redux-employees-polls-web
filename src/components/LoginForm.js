/***************************************************************************************
 *    Login form  ONLY html css is somewhat copied
 *    Title: LOGIN FORM
 *    Author: Mohithpoojary
 *    Made With: HTML / CSS
 *    Availability: https://codepen.io/Mohuth/pen/QWgrPvp
 ***************************************************************************************/

import { useState } from "react";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";

const LoginForm = (props) => {
	const [username, setUsername] = useState("");
	const [password, setPassword] = useState("");
	const [errorMessage, setErrorMessage] = useState(false);
	const [successMessage, setSuccessMessage] = useState(false);

	const handleUsernameChange = (e) => {
		setUsername(e.target.value);
	};

	const handlePasswordChange = (e) => {
		setPassword(e.target.value);
	};

	const handleSubmit = (e) => {
		e.preventDefault();
		const user = props.users[username];

		if (user && user.password === password) {
			setErrorMessage(false);
			setSuccessMessage(true);
			props.dispatch(setAuthedUser(user.id));
			setUsername("");
			setPassword("");
		} else {
			setErrorMessage(true);
			setSuccessMessage(false);
		}
	};

	return (
		<div className="container">
			<div className="screen">
				<h1>Login to see Polls</h1>
				{errorMessage && (
					<h1 data-testid="error-header">
						The Login Credentials are wrong. Please try Again
					</h1>
				)}
				{successMessage && (
					<h1 data-testid="success-header">LogIn Success!</h1>
				)}
				<form className="login" onSubmit={handleSubmit}>
					<div className="login__field">
						<input
							data-testid="username"
							type="text"
							value={username}
							className="login__input"
							placeholder="User name / Email"
							onChange={handleUsernameChange}
						/>
					</div>
					<div className="login__field">
						<input
							data-testid="password"
							type="password"
							value={password}
							className="login__input"
							placeholder="Password"
							onChange={handlePasswordChange}
						/>
					</div>
					<button
                        data-testid="submit-btn"
						className="button login__submit"
						type="submit"
						disabled={username === "" || password === ""}
					>
						<span className="button__text">Log In Now</span>
					</button>
				</form>
			</div>
		</div>
	);
};

const mapStateToProps = ({ users }) => ({
	users,
});

export default connect(mapStateToProps)(LoginForm);
