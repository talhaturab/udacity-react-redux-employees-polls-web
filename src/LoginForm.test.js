import { render, screen, fireEvent } from "@testing-library/react";
import * as React from "react";
import LoginForm from "./components/LoginForm";
import { createStore } from "redux";
import { Provider } from "react-redux";
import reducer from "./reducers/index";
import middleware from "./middleware/index";
import "@testing-library/jest-dom";

describe("LoginForm Snapshot", () => {
	it("matches the snapshot of the Login page", () => {
		const store = createStore(reducer, middleware);
		const view = render(
			<Provider store={store}>
				<LoginForm />
			</Provider>
		);
		expect(view).toMatchSnapshot();
	});
});

describe("Login Field exist", () => {
	it("test if the login page has all the correct fields", () => {
		const store = createStore(reducer, middleware);
		render(
			<Provider store={store}>
				<LoginForm />
			</Provider>
		);
		const logInHeading = screen.getAllByText("Login to see Polls");
		expect(logInHeading.length).toEqual(1);

		expect(screen.getByTestId("username")).toBeInTheDocument();

		expect(screen.getByTestId("password")).toBeInTheDocument();
		expect(screen.queryByTestId("success-header")).not.toBeInTheDocument();
		expect(screen.queryByTestId("error-header")).not.toBeInTheDocument();
	});
});

describe("Login Fail Test", () => {
	it("test to check login fails without correct information", () => {
		const store = createStore(reducer, middleware);
		render(
			<Provider store={store}>
				<LoginForm />
			</Provider>
		);
		const usernameInput = screen.getByTestId("username");
		fireEvent.change(usernameInput, { target: { value: "wrongUsername" } });

        const passwordInput = screen.getByTestId("password");
		fireEvent.change(passwordInput, { target: { value: "wrongpassword" } });

        const submitButton = screen.getByTestId("submit-btn");
        fireEvent.click(submitButton);

        expect(screen.getByTestId('error-header')).toBeInTheDocument();
        expect(screen.queryByTestId('success-header')).not.toBeInTheDocument();
	});
});

describe("Login Button Test", () => {
	it("Login button is disabled when username field is not filled", () => {
		const store = createStore(reducer, middleware);
		render(
			<Provider store={store}>
				<LoginForm />
			</Provider>
		);

        const passwordInput = screen.getByTestId("password");
		fireEvent.change(passwordInput, { target: { value: "password" } });

        const submitButton = screen.getByTestId("submit-btn");
        expect(submitButton).toBeDisabled()
	});

    it("Login button is disabled when password field is not filled", () => {
		const store = createStore(reducer, middleware);
		render(
			<Provider store={store}>
				<LoginForm />
			</Provider>
		);

        const usernameInput = screen.getByTestId("username");
		fireEvent.change(usernameInput, { target: { value: "username" } });

        const submitButton = screen.getByTestId("submit-btn");
        expect(submitButton).toBeDisabled()
	});
});
