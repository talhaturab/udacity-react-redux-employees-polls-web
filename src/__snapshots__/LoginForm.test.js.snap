// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Snapshot matches the snapshot of the Login page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="screen"
        >
          <h1>
            Login to see Polls
          </h1>
          <form
            class="login"
          >
            <div
              class="login__field"
            >
              <input
                class="login__input"
                data-testid="username"
                placeholder="User name / Email"
                type="text"
                value=""
              />
            </div>
            <div
              class="login__field"
            >
              <input
                class="login__input"
                data-testid="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <button
              class="button login__submit"
              data-testid="submit-btn"
              disabled=""
              type="submit"
            >
              <span
                class="button__text"
              >
                Log In Now
              </span>
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="screen"
      >
        <h1>
          Login to see Polls
        </h1>
        <form
          class="login"
        >
          <div
            class="login__field"
          >
            <input
              class="login__input"
              data-testid="username"
              placeholder="User name / Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="login__field"
          >
            <input
              class="login__input"
              data-testid="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <button
            class="button login__submit"
            data-testid="submit-btn"
            disabled=""
            type="submit"
          >
            <span
              class="button__text"
            >
              Log In Now
            </span>
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
